File: Simulation/Car_c.py
# -*- coding: utf-8 -*-

import numpy as np
import math
from numba import jit
import matplotlib.pyplot as plt

class Car_c:
    """ Class containing model of Car """
    
    OutputFolder = ''
    
    #Constants
    Mass = 0  #kg
    
    WheelDiameter = 0  #m
    WhellInertia = 0 #kgm^2
    
    GearRatio = 0
    GearEfficiency = 0 #fraction
    GearInertia = 0
    
    RollingResistanceCoefficient = 0
    TireDrag = 0
    
    BearingDragCoefficient = 0 #N
    BearingBoreDiameter = 0
    BearingDrag = 0    
    
    AreodynamicDragCoefficient = 0
    AirDrag = 0
    
    FrontalArea = 0 #m^2
    
    DataPoints = ''
    SimulationTime = ''
    TimeInterval = ''
    
    #Variable arrays
    Acceleration = ''
    Speed = ''
    DistanceTravelled = ''
    AirDrag = ''    
    Milage = ''    
    
    #equations
    
        
    def __init__(self,SimulationTime,TimeInterval,OutputFolder):
        #class constructor
        print('Car Object Created')
        
        self.OutputFolder = OutputFolder
        
        self.SimulationTime = SimulationTime
        self.Timeinterval = TimeInterval
        self.DataPoints = math.floor(SimulationTime/TimeInterval)         
        
        #make time array for plotting
        self.TimeEllapsed = np.arange(self.DataPoints)*TimeInterval
        
        #Allocate RAM
        self.Acceleration = np.zeros(self.DataPoints)
        self.Speed = np.zeros(self.DataPoints)
        self.DistanceTravelled = np.zeros( self.DataPoints )
        self.AirDrag = np.zeros( self.DataPoints )

    @jit
    def calc_AirDrag(self,AirDensity,Speed):
        AirDrag = 0.5*self.AreodynamicDragCoefficient*self.FrontalArea*AirDensity*math.pow(Speed,2)
        return(AirDrag)
    
    @jit
    def calc_BearingDrag(self,BearingDragCoefficient,Mass,BearingBore,WheelDiameter):
        BearingDrag = BearingDragCoefficient*Mass*9.81*BearingBore/WheelDiameter
        return(BearingDrag)
    
    @jit
    def calc_TireDrag(self,RollingResistanceCoefficient,Mass):
        TireDrag = RollingResistanceCoefficient*Mass*9.81
        return(TireDrag)
  
    ## Plotting ##
    def plot_DistanceTime(self):
        plt.figure()
        plt.plot(self.TimeEllapsed, self.DistanceTravelled)
        plt.xlabel('Time')
        plt.ylabel('Distance (m)')
        plt.title('Car')
        plt.show()
        plt.savefig(self.OutputFolder + '\\' + 'CarDistanceTime.png')
        
    def plot_SpeedTime(self):
        plt.figure()
        plt.plot(self.TimeEllapsed, self.Speed*3.6)
        plt.xlabel('Time')
        plt.ylabel('Speed (km.h)')
        plt.title('Car')
        plt.show()
        plt.savefig(self.OutputFolder + '\\' + 'CarSpeedTime.png')
        
    def plot_AccelerationTime(self):
        plt.figure()
        plt.plot(self.TimeEllapsed, self.Acceleration)
        plt.xlabel('Time')
        plt.ylabel('Acceleration (m/s2)')
        plt.title('Car')
        plt.show()
        plt.savefig(self.OutputFolder + '\\' + 'CarAccelerationTime.png')
        
    def plot_Milage(self):
        plt.figure()
        plt.plot(self.TimeEllapsed, self.InstantaneousMilage)
        plt.xlabel('Time')
        plt.ylabel('Milage km/kwh')
        plt.title('Instantaneous Milage')
        plt.show()
        plt.plot(self.TimeEllapsed, self.AverageMilage)
        plt.xlabel('Time')
        plt.ylabel('Milage km/kwh')
        plt.title('Average Milage')
        plt.show()
        plt.savefig(self.OutputFolder + '\\' + 'CarMilage.png')
        
    def plot_Drag(self):
        plt.figure()
        plt.plot(self.TimeEllapsed,self.AirDrag)
        plt.xlabel('Time (S)')
        plt.ylabel('Drag (N)')
        plt.title('Air Drag')
        plt.show()
        plt.savefig(self.OutputFolder + '\\' + 'CarDrag.png')

File: Simulation/Car_c.m
classdef Car_c < handle %goofy matlab class inheritance
    %% Class containing model of Car %%
    
    properties
        OutputFolder = '';

        %Constants
        Mass = 0;  %kg

        WheelDiameter = 0;  %m
        WhellInertia = 0; %kgm^2

        GearRatio = 0;
        GearEfficiency = 0; %fraction
        GearInertia = 0;

        RollingResistanceCoefficient = 0;
        TireDrag = 0;

        BearingDragCoefficient = 0; %N
        BearingBoreDiameter = 0;
        BearingDrag = 0;    

        AreodynamicDragCoefficient = 0;
        AirDrag = 0;

        FrontalArea = 0; %m^2

        DataPoints = '';
        SimulationTime = '';
        TimeInterval = '';
        TimeEllapsed = [];

        %Variable arrays
        Acceleration = '';
        Speed = '';
        DistanceTravelled = '';
        Milage = '';    
        
        AverageMilage = [];
        InstantaneousMilage = [];
    end

    methods
        function obj = Car_c(SimulationTime,TimeInterval,OutputFolder)
            %classfunction constructor
            disp('Car Object Created')

            obj.OutputFolder = OutputFolder;

            obj.SimulationTime = SimulationTime;
            obj.TimeInterval = TimeInterval;
            obj.DataPoints = floor(SimulationTime/TimeInterval);         

            %make time array for plotting
            obj.TimeEllapsed = (1:obj.DataPoints)*TimeInterval';

            %Allocate RAM
            obj.Acceleration = zeros(obj.DataPoints,1);
            obj.Speed = zeros(obj.DataPoints,1);
            obj.DistanceTravelled = zeros( obj.DataPoints,1);
            obj.AirDrag = zeros( obj.DataPoints,1);
        end

        
        function [AirDrag] = calc_AirDrag(obj,AirDensity,Speed)
            AirDrag = 0.5*obj.AreodynamicDragCoefficient*obj.FrontalArea*AirDensity*Speed^2;
        end

        
        function [BearingDrag] = calc_BearingDrag(~,BearingDragCoefficient,Mass,BearingBore,WheelDiameter)
            BearingDrag = BearingDragCoefficient*Mass*9.81*BearingBore/WheelDiameter;
        end

        
        function [TireDrag] = calc_TireDrag(~,RollingResistanceCoefficient,Mass)
            TireDrag = RollingResistanceCoefficient*Mass*9.81;
        end

        %% Plotting %%
        function plot_DistanceTime(obj,savef)
            figure()
            plot(obj.TimeEllapsed, obj.DistanceTravelled)
            xlabel('Time')
            ylabel('Distance (m)')
            title('Car')            
            if savef
                savefig([obj.OutputFolder '\\' 'CarDistanceTime.fig'])
            end
            saveas(gcf,[obj.OutputFolder '\\' 'CarDistanceTime.png'])
            close
        end

        function plot_SpeedTime(obj,savef)
            figure()
            plot(obj.TimeEllapsed, obj.Speed*3.6)
            xlabel('Time')
            ylabel('Speed (km.h)')
            title('Car')
            if savef
                savefig([obj.OutputFolder '\\' 'CarSpeedTime.fig'])
            end
            saveas(gcf,[obj.OutputFolder '\\' 'CarSpeedTime.png'])
            close
        end

        function plot_AccelerationTime(obj,savef)
            figure()
            plot(obj.TimeEllapsed, obj.Acceleration)
            xlabel('Time')
            ylabel('Acceleration (m/s2)')
            title('Car')
            if savef
                savefig([obj.OutputFolder '\\' 'CarAccelerationTime.fig'])
            end
            saveas(gcf,[obj.OutputFolder '\\' 'CarAccelerationTime.png'])
            close
        end

        function plot_Milage(obj,savef)
            figure()
            plot(obj.TimeEllapsed, obj.InstantaneousMilage)
            xlabel('Time')
            ylabel('Milage km/kwh')
            title('Instantaneous Milage')
            hold on 
            plot(obj.TimeEllapsed, obj.AverageMilage)
            legend('InstantaneousMilage','Average Milage')
            if savef
                savefig([obj.OutputFolder '\\' 'CarMilage.fig'])
            end
            saveas(gcf,[obj.OutputFolder '\\' 'CarMilage.png'])
            close
        end

        function plot_Drag(obj,savef)
            figure()
            plot(obj.TimeEllapsed,obj.AirDrag)
            hold on
            plot(obj.TimeEllapsed,ones(size(obj.TimeEllapsed))*obj.BearingDrag)
            plot(obj.TimeEllapsed,ones(size(obj.TimeEllapsed))*obj.TireDrag)
            xlabel('Time (S)')
            ylabel('Drag (N)')
            title('Drag Forces')
            legend('AirDrag','BearingDrag','TireDrag')
            if savef
                savefig([obj.OutputFolder '\\' 'CarDrag.fig'])
            end
            saveas(gcf,[obj.OutputFolder '\\' 'CarDrag.png'])
            close
        end
    end
end